kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java

```diff
@@ -6,109 +6,95 @@
+     *     <li>All references can be retrieved from kubernetes (we have no dangling references)</li>
+     *     <li>No referent has a ResolvedRefs=False condition, which declares that it has dangling references</li>
      * </ol>
+     * @return true if fully resolved
      */
-    public boolean isFullyResolved() {
-        return danglingReferences.isEmpty() && resourcesWithResolvedRefsFalse.isEmpty();
+    public boolean allReferentsFullyResolved() {
+        return allDanglingReferences().findAny().isEmpty() && allResolvedReferents()
+                .noneMatch(ResourcesUtil::hasResolvedRefsFalseCondition);
     }
 
     /**
-     * A result is fully reconciled if all referents have been reconciled.
-     * @return true if all referents have been reconciled
+     * @return a stream of all dangling references
      */
-    public boolean allReferentsHaveFreshStatus() {
-        return referentsWithStaleStatus.isEmpty();
-    }
-
-    public boolean anyDependenciesNotFoundFor(LocalRef<?> fromResource) {
-        return danglingReferences.stream().anyMatch(u -> u.from().equals(fromResource));
+    public Stream<DanglingReference> allDanglingReferences() {
+        return allResolutionResults().filter(ResolutionResult::dangling)
+                .map(result -> new DanglingReference(result.referrer(), result.reference()));
     }
 
-    public boolean anyResolvedRefsConditionsFalse() {
-        return !resourcesWithResolvedRefsFalse.isEmpty();
-    }
-
-    // using wildcards intentionally
-    @SuppressWarnings("java:S1452")
-    public Stream<LocalRef<?>> findResourcesWithResolvedRefsFalse() {
-        return resourcesWithResolvedRefsFalse.stream();
-    }
-
-    // using wildcards intentionally
-    @SuppressWarnings("java:S1452")
-    public Stream<LocalRef<?>> findReferentsWithStaleStatus() {
-        return referentsWithStaleStatus.stream();
-    }
-
-    // using wildcards intentionally
-    @SuppressWarnings("java:S1452")
-    public Stream<LocalRef<?>> findResourcesWithResolvedRefsFalse(String kind) {
-        return findResourcesWithResolvedRefsFalse().filter(r -> kind.equals(r.getKind()));
-    }
-
-    ClusterResolutionResult addAllResourcesHavingResolvedRefsFalse(Set<LocalRef<?>> resolvedRefsFalse) {
-        return new ClusterResolutionResult(cluster, danglingReferences, Stream.concat(resourcesWithResolvedRefsFalse.stream(), resolvedRefsFalse.stream()).collect(
-                Collectors.toSet()), referentsWithStaleStatus);
+    /**
+     * @return a stream of all resolved referents of this Cluster, note that KafkaProxy is not considered a Referent of VirtualKafkaCluster to prevent cycles
+     */
+    public Stream<HasMetadata> allResolvedReferents() {
+        return allResolutionResults()
+                .filter(resolutionResult -> !Objects.equals(resolutionResult.reference().getKind(), HasMetadata.getKind(KafkaProxy.class)))
```

Comment:
This is the ugly code I mentioned in standup, `allResolutionResults` contains results for anything that refers to a KafkaProxy, so we can find dangling refs where we could not resolve a KafkaProxy.

Until now we have not considered the staleness of the KafkaProxy resources, I was unsure if we should require a fresh status on all KafkaProxy referents in case we create a deadlock where both the KafkaProxy requires a fresh status on the VKC, and the VKC requires a fresh status on the KafkaProxy.

Reply:
would suggest we merge without an answer, keeping the current behaviour, while we discuss

Reply:
Setting aside the question of wether we should ignore Staleness of KafkaProxy for a moment. It feels strange that this type specific behaviour is buried in a generic class. Personally I think we should overload with a version that accepts the filter and ask the VKC Reconciler to pass its special sauce predicate. The default predicate would just accept everything. 

Should the KafkaProxy Reconciler have the reverse condition to break the cycle on that side too?

Reply:
I think we can wave this away by chanting "eventual consistency" can't we? By which I mean we only care about staleness for resources which have refs of some kind. A `KafkaProxy` does not have any refs, so the KPRec shouldn't care that a VKC which references the KP is stale (KPRec should ignore the staleness and proceed as if if the VKC is fresh). That's because we'd expect the VKCRec to shortly reconcile the VKC, at which point the KP should be reconciled again.

Maybe that's what you were describing when I interrupted during standup, in which case sorry if I derailed things.

Reply:
> VKC Reconciler to pass its special sauce predicate. The default predicate would just accept everything.

Both reconcilers do the same resolution work per VKC, resolving proxyRef on VKC and ingresses. It could be possible during KafkaProxy reconcilation that we observe an ingress that refers to a non-existent KafkaProxy and that should cause some failure. The difference is the source of the KafkaProxy, VKC Reconciler gets it from the JOSDK Context, KPReconciler just passes through the primary.

Both reconcilers currently want to exclude these resolved KafkaProxies from staleness checking

Reply:
@SamBarker is that a blocker for you? Maybe we could use a better term than `allReferents` maybe it's more accurately the `directReferents` of the cluster? Or we should't offer this stream of all referents munged together and clients should be forced to operate on the referents they are interested in. 

The hassle with that is that there are several places we just want to check that everything is resolved and we are able to proceed, so a bit of duplication required, or a shared static method, and at that point why not have the method on this result class.

Reply:
No not a blocker for me.

Definitely a clean up we can deal with separately.